[CodingStandard] Add DocBlockLineLengthFixer

matthiasnoback opened the pull request:
This continues @TomasVotruba 's work in #2879 

Things I'd still like to add:

- [x] If a doc block description contains multiple paragraphs, they should be word-wrapped separately, not together.

But I'd love some feedback as well. Also, do you test this on real projects? It will be interesting to find out how it works in the real world.

~I have at least one concern: this new fixer looks like it's just adding more behavior to the existing `LineLengthFixer`. Shouldn't the two be merged? Not sure myself.~ As Tomas writes: "Line lenght is sensitive topic, so it's better to have it opt-in as a standalone rule."

TomasVotruba commented:
:+1: 

> I have at least one concern: this new fixer looks like it's just adding more behavior to the existing LineLengthFixer. Shouldn't the two be merged? Not sure myself.

Line lenght is sensitive topic, so it's better to have it opt-in. I would be suprised when sudently all my docblocks would change :D

TomasVotruba commented:
> Also, do you test this on real projects? It will be interesting to find out how it works in the real world.

You can add it here to `ecs.php`

TomasVotruba commented on packages/easy-coding-standard/packages/sniff-runner/src/ValueObject/File.php:
```diff
@@ -191,9 +190,7 @@ public function addError($error, $stackPtr, $code, $data = [], $severity = 0, $f
-     * Allow only specific classes
-     *
-     * {@inheritdoc}
+     * Allow only specific classes  {@inheritdoc}
```
`{@Inheritdoc}` should be on different line

matthiasnoback replied:
Another case where the check isn't good enough maybe; line starting with `'{@'` should also not be touched.

TomasVotruba reviewed (commented):
Looks as a good start. We'll have to change few old writing-docblock-habits, as now there is a standard rule to guide them, but I think it should be doable to automate :)

matthiasnoback commented on packages/easy-coding-standard/packages/sniff-runner/src/ValueObject/File.php:
```diff
@@ -119,7 +119,8 @@ public function __construct(
 
     /**
-     * Mimics @see https://github.com/squizlabs/PHP_CodeSniffer/blob/e4da24f399d71d1077f93114a72e305286020415/src/Files/File.php#L310
+     * Mimics @see
+     * https://github.com/squizlabs/PHP_CodeSniffer/blob/e4da24f399d71d1077f93114a72e305286020415/src/Files/File.php#L310
```
If the line would start with `@` then the fixer would already leave it alone. Should this be "if the line contains '@'"? That might produce false positives.

matthiasnoback commented:
I also noticed some PHPStan-reported issues, I'll fix them as well.

matthiasnoback commented:
I've added a few extra things, based on your feedback and what I had already planned.
I've also fixed any CS and PHPStan-reported issue. Nice trick about the regex101.com links, but now you also need a PHPStan rule that ensures the link shows the current regex (kidding, not kidding ;))

matthiasnoback commented:
I'm rebasing this, but I noticed I accidentally committed the results of running the fixer on the entire code base. Oops!

matthiasnoback commented:
I have rebased this, but I'm not sure why the analysis fails. It seems like the fixer actually runs on the entire code base?

TomasVotruba commented:
You're right. These 2 lines seems testing something else:
https://github.com/symplify/symplify/blob/2e093b9338cdf87593a2977e3684c9cb84da2e30/.github/workflows/code_analysis.yaml#L45-L46

You can drop them, it should be ok then

TomasVotruba commented:
I've done through test fixtures and it looks good :+1: 

I think it's ready to merge

matthiasnoback commented:
Okay, just removed those lines. Thanks for looking into it!

matthiasnoback commented:
Oops, closed it by accident.

TomasVotruba commented:
No problem.

GitHub has weird UX solution, that you comment and then it moves the close button next to reply :D

TomasVotruba commented:
Thank you :clap:

matthiasnoback commented:
No, thank you!
